### ISSUE TITLE
HashRangesKnown task hangs despite all nodes connecting

### ISSUE THREAD

**Issue by joseph-wakeling-sociomantic** [sociomantic-tsunami/dhtproto#177]
The `HashRangesKnown` task is supposed to block until at least the specified number of nodes are connected.  However, when trying this out with an app being ported to neo, I found that it seems to block indefinitely despite all nodes being connected.

The code below outlines how the task is used: in order to support operating with partial handshake, the app first awaits completion of a `HashRangesKnown(1)` instance (ensuring that at least one node hash range has been obtained), before proceeding with an `awaitOrTimeout` on all hash ranges being known.

```D
// wait for dht.neo protocol to obtain all hash node ranges
log.info("Waiting for DHT.neo nodes hash ranges ...");
theScheduler.await(dht.blocking.new HashRangesKnown(1));

log.info("At least one DHT.neo hash range known");

bool neo_handshake_timed_out = false;
auto neo_timeout_sec = 60;

if (neo_timeout_sec > 0)
{
    neo_handshake_timed_out =
        theScheduler.awaitOrTimeout(
            dht.blocking.new AllHashRangesKnown,
            neo_timeout_sec * 1_000_000 /* microsec */);
}
else
{
    dht.blocking.waitAllHashRangesKnown();
}

if (neo_handshake_timed_out)
{
    log.error("DHT.neo hash ranges not complete within "
              ~ "{} seconds.", neo_timeout_sec);
}
else
{
    log.info("All DHT.neo hash ranges known.");
}
```

When run, the DHT client logs seem to report connections to both the local test nodes used, but the "At least one" log message is never output and the app logic does not proceed beyond the point where `theScheduler.await` is called with the `HashRangesKnown` instance:

```
Info [...] Waiting for DHT.neo nodes hash ranges ...
Trace [...] <connected>: Node [127, 0, 0, 1]:9100
Trace [...] <connected>: Node [127, 0, 0, 1]:9101
Trace [...] <hash_range_queried>: Node [127, 0, 0, 1]:9100
Trace [...] <hash_range_queried>: Node [127, 0, 0, 1]:9101
```

**Comment by joseph-wakeling-sociomantic** [comment]
FYI using `dht.blocking.waitAllHashRangesKnown();` to wait for all hash ranges works just fine.

**Comment by gavin-norman-sociomantic** [comment]
@joseph-wakeling-sociomantic could you try this again with some extra debug printouts in https://github.com/sociomantic-tsunami/swarm/blob/v5.x.x/src/swarm/neo/client/mixins/TaskBlockingCore.d#L156-L178? (I'd suggest in the notifier and after `task.suspend`, for example.)

**Comment by gavin-norman-sociomantic** [comment]
I seem to have reproduced this locally:
```D
void main ( )
{
    class MyTask : Task
    {
        override void run ( )
        {
            auto dht = new DhtClient(&this.not);
            dht.neo.addNodes("dht.nodes");
            Stdout.formatln("Starting wait");
            theScheduler.await(dht.blocking.new HashRangesKnown(1));
            Stdout.formatln("Wait over");

            // this locks
            bool timed_out = theScheduler.awaitOrTimeout(
                dht.blocking.new AllHashRangesKnown,
                1_000_000);

            // this doesn't lock
            // dht.blocking.waitAllHashRangesKnown();

            Stdout.formatln("YES! {}", timed_out);
            theScheduler.shutdown();
        }

        private void not ( DhtClient.Neo.DhtConnNotification info )
        {
            char[] buf;
            formatNotification(info, buf);
            Stdout.formatln("Notifier: {}", buf);
        }
    }

    initScheduler(Scheduler.Configuration.init);
    theScheduler.schedule(new MyTask);
    theScheduler.eventLoop();
}
```

**Comment by joseph-wakeling-sociomantic** [comment]
Do you need me to follow up with the debug printouts, or are you OK to follow up from here?

**Comment by gavin-norman-sociomantic** [comment]
Hmm... the code above actually seems to work or fail at random :/

**Comment by joseph-wakeling-sociomantic** [comment]
OK, I'll try to follow up later today.

**Comment by gavin-norman-sociomantic** [comment]
>Do you need me to follow up with the debug printouts, or are you OK to follow up from here?

If your test case is reliably reproducible: yes please.

**Comment by gavin-norman-sociomantic** [comment]
Ok, no thanks, I think I have it now.

Failure case (both nodes connect before either hash range query returns):
```
Starting wait
Checking: 0 / 1
Notifier: <connected>: Node [127, 0, 0, 1]:10100
Checking: 0 / 1
Notifier: <connected>: Node [127, 0, 0, 1]:10101
Checking: 0 / 1
Notifier: <hash_range_queried>: Node [127, 0, 0, 1]:10100
Notifier: <hash_range_queried>: Node [127, 0, 0, 1]:10101
```
Success case (one hash range query returns before the other node connects):
```
Starting wait
Checking: 0 / 1
Notifier: <connected>: Node [127, 0, 0, 1]:10100
Checking: 0 / 1
Notifier: <hash_range_queried>: Node [127, 0, 0, 1]:10100
Notifier: <connected>: Node [127, 0, 0, 1]:10101
Checking: 1 / 1
Wait over
```

**Comment by gavin-norman-sociomantic** [comment]
Ahh, got it. The DHT client's blocking connect methods are using https://github.com/sociomantic-tsunami/swarm/blob/v5.x.x/src/swarm/neo/client/mixins/TaskBlockingCore.d#L156-L178, which only gets notified about connection events, _not about hash range events_.

**Comment by joseph-wakeling-sociomantic** [comment]
Nice catch!  I was pretty sure there must be something embuggerated about the `waitConnect` method, it seemed the only obvious place where something could be going wrong.  But I didn't think to check how the connection notifier was working.

**Comment by gavin-norman-sociomantic** [comment]
Try the PR out if you like Joe.

**Comment by joseph-wakeling-sociomantic** [comment]
This was fixed in neo-v0.8.2 (now released).

**Comment by nemanja-boric-sociomantic** [comment]
Indeed. Fixed in sociomantic-tsunami/dhtproto#178 .
